src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+import Debug "mo:base/Debug";
+import Error "mo:base/Error";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Principal "mo:base/Principal";
+import R "mo:base/Result";
+import Time "mo:base/Time";
+import Timer "mo:base/Timer";
+
+import Vec "mo:vector";
+import QueueBuffer "QueueBuffer";
+
+module {
+
+  type StreamError = { #NotRegistered; #StreamClosed };
+
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   123
+  ///   func (streamId : Nat, element: Int, index: Nat): () {
+  ///     ... do your logic with incoming item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  public class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+    startFromIndex : Nat,
+  ) {
+
+    var expectedNextIndex_ : Nat = startFromIndex;
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+
+    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > 120_000_000_000; // 2 minutes
+
+    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), { #StreamClosed }> {
+      if (isStreamClosed()) {
+        return #err(#StreamClosed);
```

Comment:
This error should include the length of the stream when it was closed, e.g. `#StreamClosed : expectedNextIndex_`.

Reply:
Done